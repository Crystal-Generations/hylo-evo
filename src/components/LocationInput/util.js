export function convertMapboxToLocation (mapboxResult) {
  const neighborhoodObject = mapboxResult.context.find(c => c.id.includes('neighborhood'))
  const postcodeObject = mapboxResult.context.find(c => c.id.includes('postcode'))
  const placeObject = mapboxResult.context.find(c => c.id.includes('place'))
  const regionObject = mapboxResult.context.find(c => c.id.includes('region'))
  const countryObject = mapboxResult.context.find(c => c.id.includes('country'))

  return {
    accuracy: mapboxResult.properties.accuracy,
    addressNumber: mapboxResult.address,
    addressStreet: mapboxResult.text,
    bbox: mapboxResult.bbox ? [{ lng: mapboxResult.bbox[0], lat: mapboxResult.bbox[1] }, { lng: mapboxResult.bbox[2], lat: mapboxResult.bbox[3] }] : [],
    center: { lng: mapboxResult.center[0], lat: mapboxResult.center[1] },
    city: placeObject && placeObject.text,
    country: countryObject && countryObject.short_code,
    fullText: mapboxResult.place_name,
    // geometry: [Point]
    // locality
    neighborhood: neighborhoodObject && neighborhoodObject.text,
    region: regionObject && regionObject.text,
    postcode: postcodeObject && postcodeObject.text
    // wikidata: String
  }
}
