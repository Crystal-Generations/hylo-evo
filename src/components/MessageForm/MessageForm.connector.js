import { connect } from 'react-redux'
import { getMe } from 'store/selectors/getMe'
import {
  getTextForMessageThread,
  createMessage,
  updateMessageText
} from './MessageForm.store'
import { CREATE_MESSAGE } from 'store/constants'

export function mapStateToProps (state, props) {
  return {
    text: getTextForMessageThread(state, props),
    pending: !!state.pending[CREATE_MESSAGE],
    currentUser: getMe(state)
  }
}

export const mapDispatchToProps = {
  createMessage,
  updateMessageText
}
export default connect(mapStateToProps, mapDispatchToProps, null, {withRef: true})
