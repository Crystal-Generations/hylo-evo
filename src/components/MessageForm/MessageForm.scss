$height: 56px;

.message-form {
  border: $light-border;
  border-radius: 4px;
  width: 100%;
  min-height: $height;
  padding: $space-2x $space-4x;
  display: flex;
  background: $color-white;
}

.message-textarea {
  composes: bdy-drk-lg from 'css/typography.scss';
  flex: 1;
  border: none;
  resize: none;
  outline: none;
  padding: 0 $space-2x;
}

.user-image {

}
