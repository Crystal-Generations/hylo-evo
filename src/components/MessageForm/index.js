import component from './MessageForm'
import connector from './connector'
export default connector(component)
