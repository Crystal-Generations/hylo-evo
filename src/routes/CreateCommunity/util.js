export const slugValidatorRegex = /^[0-9a-z\s-]+$/
