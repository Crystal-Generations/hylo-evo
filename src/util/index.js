import { filter } from 'lodash/fp'
import prettyDate from 'pretty-date'

export function bgImageStyle (url) {
  if (!url) return {}
  const escaped = url.replace(/([\(\)])/g, (match, $1) => '\\' + $1) // eslint-disable-line
  return {backgroundImage: `url(${escaped})`}
}

export function personUrl (person) {
  // placeholder
  return '/'
}

export const findChildLink = element => {
  if (element.nodeName === 'A') return element
  if (element.hasChildNodes()) {
    const mappedNodes = []
    for (var i = 0; i < element.childNodes.length; i++) {
      mappedNodes.push(findChildLink(element.childNodes[i]))
    }
    return filter(id => id, mappedNodes)[0]
  }
  return false
}

export const dispatchEvent = (el, etype) => {
  var evObj = document.createEvent('Events')
  evObj.initEvent(etype, true, false)
  el.dispatchEvent(evObj)
}

export function humanDate (date, short) {
  const isString = typeof date === 'string'
  const isValidDate = !isNaN(Number(date)) && Number(date) !== 0
  var ret = date && (isString || isValidDate)
    ? prettyDate.format(isString ? new Date(date) : date)
    : ''
  if (short) {
    ret = ret.replace(' ago', '')
  } else {
    // this workaround prevents a "React attempted to use reuse markup" error
    // which happens if the timestamp is less than 1 minute ago, because the
    // server renders "N seconds ago", but by the time React is loaded on the
    // client side, it's "N+1 seconds ago"
    let match = ret.match(/(\d+) seconds? ago/)
    if (match) {
      if (Number(match[1]) >= 50) return '1m ago'
      return 'just now'
    }
  }
  ret = ret.replace(/ minutes?/, 'm')
  .replace(/ hours?/, 'h')
  .replace(/ days?/, 'd')
  .replace(/ weeks?/, 'w')
  .replace(/ month(s?)/, ' mo$1')
  return ret
}
