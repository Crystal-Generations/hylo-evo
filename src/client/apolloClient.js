import { ApolloClient } from 'apollo-client'
import { ApolloLink } from 'apollo-link'
import HolochainWebSocketLink from './HolochainWebSocketLink'
import { RetryLink } from 'apollo-link-retry'
import { InMemoryCache } from 'apollo-cache-inmemory'
// import { HttpLink } from 'apollo-link-http'
// import { split } from 'apollo-link'
// import { get } from 'lodash/fp'
import { HOLOCHAIN_ACTIVE } from 'util/holochain'

// * FOR USING APOLLO FOR BOTH HYLO-NODE API AND HOLOCHAIN CLIENT
// const link = split(
//   operation => HOLOCHAIN_ACTIVE,
//   new HolochainWebSocketLink({
//     uri: process.env.HOLOCHAIN_WEBSOCKET_URI
//   }),
//   new HttpLink({
//     uri: 'http://localhost:9000/noo/graphql'
//   })
// )

// * HOW TO CREATE LOCAL RESOLVERS
// https://www.apollographql.com/docs/link/links/state
// const resolvers = {
//   Mutation: {},
//   Query: {}
// }

const link = ApolloLink.from([
  new RetryLink(),
  new HolochainWebSocketLink({
    uri: process.env.HOLOCHAIN_WEBSOCKET_URI,
    active: HOLOCHAIN_ACTIVE
  })
])

const apolloClient = new ApolloClient({
  link,
  // resolvers
  cache: new InMemoryCache(),
  connectToDevTools: true
})

export default apolloClient
