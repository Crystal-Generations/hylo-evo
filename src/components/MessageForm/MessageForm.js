import React from 'react'
import { throttle } from 'lodash'
import { onEnterNoShift } from 'util/textInput'
import { getSocket, socketUrl } from 'client/websockets'
import RoundImage from 'components/RoundImage'
var { func, object, string, bool } = React.PropTypes
import './MessageForm.scss'

// should share with comment form
export const STARTED_TYPING_INTERVAL = 3000

export default class MessageForm extends React.Component {
  static propTypes = {
    className: string,
    currentUser: object,
    postId: string,
    placeholder: string,
    onFocus: func,
    onBlur: func,
    pending: bool,
    createComment: func
  }

  constructor (props) {
    super(props)
    this.state = {}
  }

  submit = event => {
    if (event) event.preventDefault()
    if (!this.state.text) return false
    const { postId, createComment } = this.props
    const { currentUser } = this.context
    const userId = currentUser.id
    const { text } = this.state
    createComment({postId, text, userId})
    this.startTyping.cancel()
    this.sendIsTyping(false)
    this.setState({text: ''})
    return false
  }

  componentDidMount () {
    this.socket = getSocket()
  }

  focus () {
    this.refs.editor.focus()
  }

  isFocused () {
    return this.refs.editor === document.activeElement
  }

  sendIsTyping (isTyping) {
    const { postId } = this.props
    if (this.socket) {
      this.socket.post(socketUrl(`/noo/post/${postId}/typing`), {isTyping})
    }
  }

  // broadcast "I'm typing!" every 5 seconds starting when the user is typing.
  // We send repeated notifications to make sure that a user gets notified even
  // if they load a comment thread after someone else has already started
  // typing.
  startTyping = throttle(() => {
    this.sendIsTyping(true)
  }, STARTED_TYPING_INTERVAL)

  render () {
    const { onFocus, onBlur, className, currentUser } = this.props
    const placeholder = this.props.placeholder || 'Write something...'
    const { text } = this.state

    const handleKeyDown = e => {
      this.startTyping()
      onEnterNoShift(e => {
        e.preventDefault()
        this.submit()
      }, e)
    }

    return <form onSubmit={this.submit} styleName='message-form' className={className}>
      <RoundImage url={currentUser.avatarUrl} styleName='user-image' medium />
      <textarea ref='editor' name='message' value={text} styleName='message-textarea'
        placeholder={placeholder}
        onFocus={onFocus}
        onChange={e => this.setState({text: e.target.value})}
        onBlur={onBlur}
        onKeyUp={this.stopTyping}
        onKeyDown={handleKeyDown} />
    </form>
  }
}
