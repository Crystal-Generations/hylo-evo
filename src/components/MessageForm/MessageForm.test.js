import MessageForm from './MessageForm'
import { socketUrl } from 'client/websockets'
import { keyMap } from 'util/textInput'
import { shallow, mount } from 'enzyme'
import React from 'react'

it('to match the latest snapshot', () => {
  const currentUser = {
    id: '1',
    avatarUrl: 'http://image.com/p.png'
  }
  const wrapper = shallow(<MessageForm messageThreadId={'1'} text='hey you' currentUser={currentUser} />)
  expect(wrapper).toMatchSnapshot()
})

describe('MessageForm', () => {
  const currentUser = {
    id: '1',
    avatarUrl: 'http://image.com/p.png'
  }
  const socket = {
    post: jest.fn()
  }
  const messageThreadId = '1'
  const createMessage = jest.fn()
  const wrapper = mount(<MessageForm
    messageThreadId={messageThreadId}
    text='hey you'
    currentUser={currentUser}
    socket={socket}
    createMessage={createMessage} />)

  it('to send startTyping when typing happens', () => {
    wrapper.find('textarea').simulate('keydown')
    const expectedSocketUrl = socketUrl(`/noo/post/1/typing`)
    const expectedTypingValue = { isTyping: true }
    expect(socket.post.mock.calls[0][0]).toEqual(expectedSocketUrl)
    expect(socket.post.mock.calls[0][1]).toEqual(expectedTypingValue)
  })

  it('to not createMessage when enter with shift is pressed', () => {
    wrapper.find('textarea').simulate('keydown', {which: keyMap.ENTER, shiftKey: true})
    expect(createMessage).not.toHaveBeenCalled()
  })

  it('to createMessage when enter (no shift) is pressed', () => {
    wrapper.find('textarea').simulate('keydown', {which: keyMap.ENTER})
    expect(createMessage.mock.calls[0]).toEqual([messageThreadId, 'hey you', currentUser.id])
  })
})
