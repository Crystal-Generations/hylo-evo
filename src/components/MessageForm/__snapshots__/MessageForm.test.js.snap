// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`to match the latest snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <RoundImage
    medium={true}
    url="http://image.com/p.png"
  />
  <textarea
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Write something..."
    value="hey you"
  />
</form>
`;
